refactor(router-core): consistent Unicode handling

nlynzaad opened the pull request:
This PR follows on from #5663 and ensures that the path names are always in a consistent state by decoding the pathname when parsing the location. This ensures that the pathname is always in a decoded state and any further encoding/decoding that is required for navigation and matching is done in a consistent manner. While the pathname is decoded the URL is not ensuring the href and urls are always in a valid state.

One case where this currently fails currently is when matching a link as active or not when the params contain special characters, see #5008. While the resolved pathname are in encoded form, the params are on the link is not, and this can cause mismatching to occur.  

As an added bonus, by decoding the path upfront we do not need to decode each path segment individually when parsing the paths for matching.

This PR adds a bunch of tests to ensure this is treated consistently throughout router at the core, code-based routing in React and Solid as well as in the browser environment with file-based routing in React and Solid e2e tests.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Added support for Unicode characters (including emoji and non-Latin scripts) in route paths and parameters
  * Enhanced handling of complex URL encoding and special characters in navigation
  * New route examples demonstrating international path support

* **Performance**
  * Reduced page load delay for improved responsiveness

* **Testing**
  * Expanded test coverage for Unicode paths and encoded parameter handling
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

SeanCassiere commented on packages/react-router/tests/router.test.tsx:
```diff
@@ -374,7 +374,7 @@ describe('encoding: URL param segment for /posts/$slug', () => {
     await act(() => router.load())
 
-    expect(router.state.location.pathname).toBe('/posts/%F0%9F%9A%80')
+    expect(router.state.location.pathname).toBe('/posts/🚀')
```
Not sure what needs to be changed here.

- Expectation of the test.
- Or the test name itself (i.e `it('should ...')`).

SeanCassiere commented on packages/router-core/src/router.ts:
```diff
@@ -1173,7 +1173,7 @@ export class RouterCore<
         publicHref: href,
         url: url.href,
-        pathname,
+        pathname: decodePathSegment(pathname),
```
Maybe a function rename here, since its no longer `decodePathSegment` (which was using `decodeURIComponent`) rather its more along the lines of `decodePath` (using `decodeURI`).

nlynzaad commented on packages/react-router/tests/router.test.tsx:
```diff
@@ -623,7 +628,8 @@ describe('encoding: URL splat segment for /$', () => {
 
-    expect(router.state.location.pathname).toBe('/%F0%9F%9A%80')
+    expect(router.state.location.href).toBe('/%F0%9F%9A%80')
+    expect(router.state.location.pathname).toBe('/🚀')
```
yes, the url and href is to remain encoded. the pathname not

nlynzaad commented:
requested changes made. 

quick note on the actionable comment from code rabbit. in the specific test `@` is not added to the pathParamsAllowedCharacters and hence is encoded. This is specifically left as such and hence the use of decodeURI in decodePath.

nlynzaad commented:
> I still think that the test names need to be changed to match what we're now testing for.
> 
> Other than that, I'm OK with the changes themselves.

reworked the test names and cleaned up the structure a bit
